File: 20-actors/mizuho/methods/test_water_supply.py
"""Tests for mizuho water-supply operational loop.

    cd 20-actors/mizuho/methods
    PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 python3 -m pytest
"""

from __future__ import annotations

import pytest

from _substrate import SafetyError
from water_supply import (
    MAX_SERVICE_POPULATION,
    ReservoirPlant,
    commission_water_supply,
    to_datoms,
)


def test_supply_restores_level_after_demand_step():
    res = commission_water_supply(demand_step_lps=20.0)
    assert res.level_restored
    assert res.final_level_m == pytest.approx(3.0, abs=1e-2)  # back to service setpoint
    assert res.settling_seconds > 0
    assert res.final_pressure_bar > 0  # service pressure restored


def test_supply_restores_for_large_demand_step():
    # A bigger demand (more taps open) is also rejected back to the setpoint.
    res = commission_water_supply(demand_step_lps=80.0, service_population=1500)
    assert res.level_restored
    assert res.final_level_m == pytest.approx(3.0, abs=1e-2)


@pytest.mark.parametrize("use", ["weapon", "fire-control", "interdiction", "flood"])
def test_non_civilian_use_refused(use):
    with pytest.raises(SafetyError):
        commission_water_supply(demand_step_lps=20.0, use=use)


def test_community_scale_cap_enforced_g3():
    # A service population above the community-scale cap is N1 (a municipal
    # utility) and is structurally refused before any run.
    with pytest.raises(SafetyError):
        commission_water_supply(
            demand_step_lps=20.0, service_population=MAX_SERVICE_POPULATION + 1
        )


def test_at_cap_is_allowed():
    res = commission_water_supply(
        demand_step_lps=20.0, service_population=MAX_SERVICE_POPULATION
    )
    assert res.service_population == MAX_SERVICE_POPULATION
    assert res.level_restored


def test_reservoir_self_regulates():
    # No pump command: a gravity-fed tank with a head-dependent leak drains
    # toward a lower equilibrium (real first-order dynamics, not free fall to 0).
    tank = ReservoirPlant(area_m2=20.0, level_m=3.0, demand_lps=10.0)
    start = tank.measure()
    for _ in range(100):
        tank.step(0.0, 1.0)
    assert tank.measure() < start
    assert tank.measure() >= 0.0


def test_datoms_are_aggregate_dry_run_no_server_key():
    res = commission_water_supply(demand_step_lps=20.0)
    d = to_datoms(res, "spring-001")
    assert d[":water.supply/dry-run"] is True
    assert d[":water.supply/server-held-key"] is False
    assert d[":water.supply/representative"] is True
    assert d[":water.supply/level-restored"] is True
    assert d[":water.supply/service-population"] <= MAX_SERVICE_POPULATION

File: 20-actors/mizuho/methods/test_water_supply.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/test_water_supply.py.
(ns mizuho.methods.test-water-supply
  "Tests for mizuho water-supply operational loop.

  Run:  bb --classpath 20-actors 20-actors/mizuho/methods/test_water_supply.clj"
  (:require [mizuho.methods.substrate :as s]
            [mizuho.methods.water-supply :as ws]
            [clojure.test :refer [deftest is run-tests]]))

(defn- approx [a b tol] (<= (Math/abs (double (- a b))) tol))

(deftest supply-restores-level-after-demand-step
  (let [r (ws/commission-water-supply {:demand-step-lps 20.0})]
    (is (:level-restored r))
    (is (approx (:final-level-m r) 3.0 1e-2))   ; back to service setpoint
    (is (> (:settling-seconds r) 0))
    (is (> (:final-pressure-bar r) 0))))        ; service pressure restored

(deftest supply-restores-for-large-demand-step
  (let [r (ws/commission-water-supply {:demand-step-lps 80.0 :service-population 1500})]
    (is (:level-restored r))
    (is (approx (:final-level-m r) 3.0 1e-2))))

(deftest non-civilian-use-refused
  (doseq [use ["weapon" "fire-control" "interdiction" "flood"]]
    (is (thrown? clojure.lang.ExceptionInfo
                 (ws/commission-water-supply {:demand-step-lps 20.0 :use use}))
        (str "N1 must refuse use " use))))

(deftest community-scale-cap-enforced-g3
  ;; A service population above the community-scale cap is N1 (a municipal utility), refused.
  (is (thrown? clojure.lang.ExceptionInfo
               (ws/commission-water-supply {:demand-step-lps 20.0
                                            :service-population (inc ws/MAX-SERVICE-POPULATION)}))))

(deftest at-cap-is-allowed
  (let [r (ws/commission-water-supply {:demand-step-lps 20.0
                                       :service-population ws/MAX-SERVICE-POPULATION})]
    (is (= (:service-population r) ws/MAX-SERVICE-POPULATION))
    (is (:level-restored r))))

(deftest reservoir-self-regulates
  ;; No pump command: a gravity-fed tank with a head-dependent leak drains toward a lower
  ;; equilibrium (real first-order dynamics, not free fall to 0).
  (let [tank (ws/reservoir-plant {:area-m2 20.0 :level-m 3.0 :demand-lps 10.0})
        start ((:measure tank))]
    (dotimes [_ 100] ((:step tank) 0.0 1.0))
    (is (< ((:measure tank)) start))
    (is (>= ((:measure tank)) 0.0))))

(deftest datoms-are-aggregate-dry-run-no-server-key
  (let [d (ws/to-datoms (ws/commission-water-supply {:demand-step-lps 20.0}) "spring-001")]
    (is (true? (:water.supply/dry-run d)))
    (is (false? (:water.supply/server-held-key d)))
    (is (true? (:water.supply/representative d)))
    (is (true? (:water.supply/level-restored d)))
    (is (<= (:water.supply/service-population d) ws/MAX-SERVICE-POPULATION))))

;; the N1/G3 refusals are SafetyError-typed
(deftest refusals-are-safety-errors
  (doseq [args [{:demand-step-lps 20.0 :use "weapon"}
                {:demand-step-lps 20.0 :service-population (inc ws/MAX-SERVICE-POPULATION)}]]
    (is (try (ws/commission-water-supply args) false
             (catch clojure.lang.ExceptionInfo e (s/safety-error? e))))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'mizuho.methods.test-water-supply)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
